### PR TITLE
free_nodes optimization

### PR DESCRIPTION
**Pull request by bremanandjk**
#### Describe Bug or Feature

On every job exit, the free_nodes() is consuming more time for releasing the nodes associated with the job object. 

#### Describe Your Change

Optimized in a way that instead of the loop through to "each subnode" in "each node" in "each mom", identifying the node directly from chunks and releasing nodes using deallocate_job_from_node()

#### Attach Test and Valgrind Logs/Output

Before fix: 
Before_fix_free_nodes
After fix:
After_fix_free_nodes

Done these tests using following steps: 
Install pbs
kill running mom kill -9 $(pidof pbs_mom)
run pbs_mom -m as root to start mock mom
pbs_config --vnodify -N 10 -a resources_available.ncpus=500 (to create 10 vnodes with 500 ncpus each)
submit 5k jobs (for i in $(seq 1 1 5000); do qsub -koe -- /bin/date; done; as pbsroot user)
start only one sched cycle (qmgr -c "s s scheduling=1" && qmgr -c "s s scheduling=0")

### CODE REVIEW

**Comment by hirenvadalia** [comment]
@bremanandjk Can please you try with 50 vnodes with 1000 cpu each and submit 50k jobs then measure before/after fix.

**Comment by bremanandjk** [comment]
> @bremanandjk Can please you try with 50 vnodes with 1000 cpu each and submit 50k jobs then measure before/after fix.

50k jobs result from blrentperf04 machine

before fix:
50k_before_fix

after fix:
50k_after_fix

**Comment by hirenvadalia** [comment]
@bremanandjk Thanks for running but graphs don't look correct or I am missing something... Anyways will do more runs on whole branch so will find things at that time.

File: src/server/node_manager.c [hunk 1]
```diff
@@ -6947,125 +6982,11 @@ set_nodes(void *pobj, int objtype, char *execvnod_in, char **execvnod_out, char
...
-		}
-		special_case = 0;
-	}
-	pjob->ji_qs.ji_svrflags &= ~JOB_SVFLG_HasNodes;
+	char	*pnodespec = pjob->ji_wattr[(int)JOB_ATR_exec_vnode].at_val.at_str;
+	if (!pnodespec)
+		return;
 
-	is_called_by_job_purge = 0;
+	dealloc_hosts(pjob, pjob->ji_wattr[(int)JOB_ATR_exec_vnode].at_val.at_str);
```
Comment (nithinj) [inline src/server/node_manager.c:6989]: Why not pass pnodespec to the function then?

File: src/server/node_manager.c [hunk 2]
```diff
@@ -6947,125 +6982,11 @@ set_nodes(void *pobj, int objtype, char *execvnod_in, char **execvnod_out, char
...
-				if (pjob->ji_qs.ji_substate == JOB_SUBSTATE_PROVISION)
-					free_prov_vnode(pnode);
-			}
-
-		}
-		special_case = 0;
-	}
-	pjob->ji_qs.ji_svrflags &= ~JOB_SVFLG_HasNodes;
+	char	*pnodespec = pjob->ji_wattr[(int)JOB_ATR_exec_vnode].at_val.at_str;
+	if (!pnodespec)
```
Comment (nithinj) [inline src/server/node_manager.c:6986]: Check for SET flag

File: src/server/node_manager.c [hunk 3]
```diff
@@ -6936,6 +6936,41 @@ set_nodes(void *pobj, int objtype, char *execvnod_in, char **execvnod_out, char
 	return 0;
 }
 
+
+void
+dealloc_hosts(job  *pjob, char *execvnod_in) {
```
Comment (nithinj) [inline src/server/node_manager.c:6941]: Do we need the second argument if it can be fetched from pjob?

Comment (nithinj) [reply src/server/node_manager.c:6866]: @bremanandjk do you got a chance to work on this?

Comment (bremanandjk) [reply src/server/node_manager.c:6866]: @nithinj  I thought of removing this function and added the code in free_nodes function itself. Refer to the mainline PR.

Let me close this PR, we could cherry-pick from mainline later.

**Comment by bremanandjk** [comment]
Raised another PR that matches the changes from the mainline. Hence closing this PR.

**Pull request closed without merging**